ts/test/session/sending/MessageQueue_test.ts

```diff
@@ -0,0 +1,393 @@
+    TestUtils.stubWindow('libsignal', {
+      SignalProtocolAddress: sandbox.stub(),
+      SessionCipher: Stubs.SessionCipherStub,
+    } as any);
+
+    // Message Sender Stubs
+    sendStub = sandbox.stub(MessageSender, 'send').resolves();
+    sendToOpenGroupStub = sandbox
+      .stub(MessageSender, 'sendToOpenGroup')
+      .resolves(true);
+
+    // Group Utils Stubs
+    sandbox.stub(GroupUtils, 'isMediumGroup').returns(false);
+    groupMembersStub = sandbox
+      .stub(GroupUtils, 'getGroupMembers' as any)
+      .callsFake(async () => TestUtils.generateMemberList(10));
+
+    // Session Protocol Stubs
+    sandbox.stub(SessionProtocol, 'sendSessionRequest').resolves();
+    hasSessionStub = sandbox.stub(SessionProtocol, 'hasSession').resolves(true);
+    sendSessionRequestIfNeededStub = sandbox
+      .stub(SessionProtocol, 'sendSessionRequestIfNeeded')
+      .resolves();
+
+    // Pending Mesage Cache Stubs
+    const chatMessages = Array.from(
+      { length: 10 },
+      TestUtils.generateChatMessage
+    );
+    const rawMessage = toRawMessage(
+      TestUtils.generateFakePubkey(),
+      TestUtils.generateChatMessage()
+    );
+
+    sandbox.stub(PendingMessageCache.prototype, 'add').resolves(rawMessage);
+    sandbox.stub(PendingMessageCache.prototype, 'remove').resolves();
+    sandbox
+      .stub(PendingMessageCache.prototype, 'getDevices')
+      .returns(TestUtils.generateMemberList(10));
+    sandbox
+      .stub(PendingMessageCache.prototype, 'getForDevice')
+      .returns(
+        chatMessages.map(m => toRawMessage(TestUtils.generateFakePubkey(), m))
+      );
+
+    // Spies
+    sendSyncMessageSpy = sandbox.spy(MessageQueue.prototype, 'sendSyncMessage');
+    sendMessageToDevicesSpy = sandbox.spy(
+      MessageQueue.prototype,
+      'sendMessageToDevices'
+    );
+    sendToGroupSpy = sandbox.spy(MessageQueue.prototype, 'sendToGroup');
+
+    // Init Queue
+    messageQueueStub = new MessageQueue();
+  });
+
+  afterEach(() => {
+    TestUtils.restoreStubs();
+    sandbox.restore();
+  });
+
+  describe('send', () => {
+    it('can send to a single device', async () => {
+      const device = TestUtils.generateFakePubkey();
+      const message = TestUtils.generateChatMessage();
+
+      const promise = messageQueueStub.send(device, message);
+      await expect(promise).to.be.fulfilled;
+    });
+
+    it('can send sync message', async () => {
+      const devices = TestUtils.generateMemberList(3);
+      const message = TestUtils.generateChatMessage();
+
+      const promise = messageQueueStub.sendSyncMessage(message, devices);
+      expect(promise).to.be.fulfilled;
+    });
+  });
+
+  describe('processPending', () => {
+    it('will send sync message if no session', async () => {
+      hasSessionStub.resolves(false);
+
+      const device = TestUtils.generateFakePubkey();
+      const promise = messageQueueStub.processPending(device);
+
+      expect(promise).to.be.fulfilled;
+
+      await tick();
```

Comment:
what is the use of tick?
if it's a timeout of 0 you can use `timeout(0)` defined in test-utils.ts